### PR TITLE
ENH: Add `__deepcopy__` to MaskedConstant

### LINKED ISSUE #11021: Copy and deepcopy of np.ma.masked returns np.ma.array()
I would expect that copy or deepcopy of `np.ma.masked` would give me the same object, but this doesn't appear to be the case:
```
In [9]: copy.deepcopy(np.ma.masked)
Out[9]: 
masked_array(data=--,
             mask=True,
       fill_value=1e+20,
            dtype=float64)

In [10]: copy.copy(np.ma.masked)
Out[10]: 
masked_array(data=--,
             mask=True,
       fill_value=1e+20,
            dtype=float64)

In [11]: type(copy.deepcopy(np.ma.masked))
Out[11]: numpy.ma.core.MaskedArray

In [12]: type(np.ma.masked)
Out[12]: numpy.ma.core.MaskedConstant

In [13]: np.__version__
Out[13]: '1.14.2'
```

### PR DESCRIPTION
**Pull request by Krithi07**
Implemented deepcopy on MaskedConstant returning just self

Fixes #11021

### CODE REVIEW

**Review by eric-wieser — DISMISSED** [review]
See above, which also explains why tests fail

File: numpy/ma/core.py [hunk 1]
```diff
@@ -6317,6 +6317,9 @@ def copy(self, *args, **kwargs):
         # precedent for this with `np.bool_` scalars.
         return self
 
+	def __deepcopy__(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (Krithi07) [reply numpy/ma/core.py]: def __deepcopy__(self, memo=None):
	    return self

Does this look like the right function?

[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/ma/core.py]: I don't think you need the `=None` there, the argument is required. Otherwise, yes.

**Comment by eric-wieser** [comment]
Still needs a test of `copy.copy(np.ma.masked)` and `copy.deepcopy(np.ma.masked)` in the file I mention aboev

**Comment by Krithi07** [comment]
`def test__copy__(self):
        assert_equal(
            copy.copy(np.ma.masked) is np.ma.masked,
            copy.copy(np.True_) is np.True_)

    def test__deepcopy__(self):
        assert_equal(
            copy.deepcopy(np.ma.masked) is np.ma.masked,
            copy.deepcopy(np.True_) is np.True_)`

Do they seem alright? I have never written a test function before.

File: numpy/ma/tests/test_core.py [hunk 2]
```diff
@@ -4826,6 +4826,16 @@ def test_copy(self):
             np.ma.masked.copy() is np.ma.masked,
             np.True_.copy() is np.True_)
 
+    def test__copy(self):
+        assert_equal(
+            copy.copy(np.ma.masked) is np.ma.masked,
+            copy.copy(np.True_) is np.True_)
```
Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: Not incorrect, but I'd probably go for the simpler
```
assert_(copy.copy(np.ma.masked) is np.ma.masked)
```
(note the underscore)

File: numpy/ma/tests/test_core.py [hunk 3]
```diff
@@ -4826,6 +4826,16 @@ def test_copy(self):
...
+
+    def test__deepcopy(self):
```
Comment (eric-wieser) [inline numpy/ma/tests/test_core.py]: I'd replace this with just `test_deepcopy`, dropping the extra underscore

Comment (Krithi07) [reply numpy/ma/tests/test_core.py]: I'll make the changes and commit

Comment (Krithi07) [reply numpy/ma/tests/test_core.py]: Using the online editor to make these changes is alright?

Comment (eric-wieser) [reply numpy/ma/tests/test_core.py]: Yep, online editor is fine - I can just squash the commits together when I merge.

For bigger patches that make sense as multiple commits, it's better to squash by hand offline, but there's no need to here.

File: numpy/ma/tests/test_core.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/ma/tests/test_core.py]: I don't believe this - the code you copied this from was code that I wrote, and I disagreed with it!

**Comment by Krithi07** [comment]
Got CI error previously in the test_core.py for not having included the import statement for copy. Added `import copy` in both the added test functions.

File: numpy/ma/core.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Krithi07) [reply numpy/ma/core.py]: For the below test function to work, do you think I will have to have `=None` in the __copy__ funciton?

`def test__copy(self):
    assert_equal(
        copy.copy(np.ma.masked) is np.ma.masked,
        copy.copy(np.True_) is np.True_)`

[the remark this answers is not in the document]
Comment (Krithi07) [reply numpy/ma/core.py]: If adding none seems fine, both the commits made on 5th May can be used

File: numpy/ma/core.py [hunk 4]
```diff
@@ -6317,6 +6317,12 @@ def copy(self, *args, **kwargs):
         # precedent for this with `np.bool_` scalars.
         return self
 
+    def __copy__(self, memo=None):
```
Comment (eric-wieser) [inline numpy/ma/core.py]: `__copy__` does not take a `memo` argument - only `__deepcopy__` does

File: numpy/ma/core.py [hunk 5]
```diff
@@ -6317,6 +6317,12 @@ def copy(self, *args, **kwargs):
...
+        return self
+		
+    def __deepcopy__(self, memo=None):
```
Comment (eric-wieser) [inline numpy/ma/core.py]: I don't think you need the `=None` here

Comment (Krithi07) [reply numpy/ma/core.py]: Fixed 'em both

**Comment by Krithi07** [comment]
Hey, is there anything else that needs fixing in this pull request?

**Comment by eric-wieser** [comment]
Could do with a release note under "compatibility" - something like:
```
``copy.copy`` and ``copy.deepcopy`` no longer turn ``masked`` into an array
----------------------------------------------------------------------------
Since ``np.ma.masked`` is a readonly scalar, copying should be a no-op. These
functions now behave consistently with ``np.copy()``.
```

**Comment by eric-wieser** [comment]
Thanks @Krithi07!

**Comment by Krithi07** [comment]
Thanks @eric-wieser for guiding me through my first PR. I know I've been a pain but I will be now able to handle my PRs myself.
Thanks!

**Pull request merged**